bindings/rust/s2n-tls/src/client_hello.rs

```diff
@@ -0,0 +1,392 @@
+const MD5_HASH_SIZE: u32 = 16;
+
+impl From<FingerprintType> for s2n_tls_sys::s2n_fingerprint_type::Type {
+    fn from(value: FingerprintType) -> Self {
+        match value {
+            FingerprintType::JA3 => s2n_tls_sys::s2n_fingerprint_type::FINGERPRINT_JA3,
+        }
+    }
+}
+
+// ClientHello is an opaque wrapper struct around `s2n_client_hello`. Note that
+// the size of this type is not known, and as such it can only be used through
+// references and pointers.
+//
+// This implementation is motivated by the different memory_management required
+// for different s2n_client_hello pointers. `s2n_client_hello_parse_message`
+// returns a `*mut s2n_client_hello` which owns it's own data. This neatly fits
+//  the "smart pointer" pattern and can be represented as a `Box<T>`.
+//
+// `s2n_connection_get_client_hello` returns a `*mut s2n_client_hello` which
+// references memory owned by the connection, and therefore must not outlive
+// the connection struct. This is best represented as a reference tied to the
+// lifetime of the `Connection` struct.
+
+/// ```no_run
+/// use s2n_tls::client_hello::{ClientHello, FingerprintType};
+/// use s2n_tls::connection::Connection;
+/// use s2n_tls::enums::Mode;
+///
+/// let mut conn = Connection::new(Mode::Server);
+/// // handshake happens
+/// let mut client_hello: &ClientHello = conn.client_hello().unwrap();
+/// let mut hash = Vec::new();
+/// let string_size = client_hello.fingerprint_hash(FingerprintType::JA3, &mut hash).unwrap();
+/// // hash has been resized so that it can store the fingerprint hash
+///
+/// let mut string = String::with_capacity(string_size as usize);
+/// // string will not be resized, and the method will fail with
+/// // ErrorType::UsageError if the string doesn't have enough capacity
+/// client_hello.fingerprint_string(FingerprintType::JA3, &mut string).unwrap();
+/// ```
+pub struct ClientHello(s2n_client_hello);
+
+// safety justifications
+// 1 - casting *s2n_client_hello <-> *ClientHello: For repr(Rust),
+// repr(packed(N)), repr(align(N)), and repr(C) structs: if all fields of a
+// struct have size 0, then the struct has size 0.
+// https://rust-lang.github.io/unsafe-code-guidelines/layout/structs-and-tuples.html#zero-sized-structs
+// 2 - casting *const s2n_client_hello -> *mut s2n_client_hello: This is safe as
+// long as the data is not actually mutated. As authors of s2n-tls, we know that
+// the get_hash and get_fingerprint methods do not mutate the data, and use mut
+// pointers as a matter of convention because it makes working with s2n_stuffers
+// and s2n_blobs easier.
+
+impl ClientHello {
+    pub fn parse_client_hello(hello: &[u8]) -> Result<Box<Self>, crate::error::Error> {
+        crate::init::init();
+        let handle = unsafe {
+            s2n_client_hello_parse_message(hello.as_ptr(), hello.len() as u32).into_result()?
+        };
+        let client_hello = handle.as_ptr() as *mut ClientHello;
```

Comment:
Is this cast required? Maybe this is more normal in Rust than it looks, but your Safety comments makes it seem like maybe not. Could you just do `type ClientHello = s2n_client_hello` to alias s2n_client_hello to a more Rust name instead? Since it seems like this logic would already prevent us from adding more fields to ClientHello in the future.

Reply:
> The rust compiler/bindings (like the public s2n.h api) treat s2n_client_hello as an opaque type. E.g. there isn't any way to directly instantiate an s2n_client_hello, and it can only be referred to through an indirection like a pointer or a reference

I didn't mean add more fields to the C s2n_client_hello. I more mean the case where we need to store additional information alongside the pointer to the C s2n_client_hello, like we do for s2n_connection and s2n_config.

Reply:
Ya, we wouldn't be able to add other fields to the struct with this approach. Do you have any ideas about the sort of functionality that we might need to add in the future? 

Also I might be looking at the wrong place, but I don't think `s2n_connection` and `s2n_config` store any extra information. The structs looks pretty austere 😄 

https://github.com/aws/s2n-tls/blob/1c26d11b96cb709d945aea860f846fc30b080afa/bindings/rust/s2n-tls/src/connection.rs#L37-L39

https://github.com/aws/s2n-tls/blob/1c26d11b96cb709d945aea860f846fc30b080afa/bindings/rust/s2n-tls/src/config.rs#L19

I talk about different approaches in the "Alternatives Considered" section of the PR description, but the short recap is that the other solution that achieves both 
1. nice memory management, 
2. respects the lifetime of the client hello associated with a connection
feels significantly more complex, so I'd still vote to go with this method.

Reply:
Ah, I was thinking about the contexts we store pointers to. So we don't put the extra state on the Rust struct itself. And if we're not already doing that, that does make this less concerning.